File: summer2019/week4/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        return recursive(head);
    }
    
    public ListNode recursive(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode remaining = recursive(head.next);
        head.next.next = head;
        head.next = null;
        return remaining;
    }
}

File: 2026/week4/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current = head
        prev_node = None
        while current:
            next_node = current.next
            current.next = prev_node
            prev_node = current
            current = next_node
        return prev_node

        